### PR TITLE
 Constant propagator: improve GOTO condition propagation

### PR DESCRIPTION
**Pull request by smowton**
This improves two-way propagation to also account for many other common patterns that imply an equality (e.g. `x != y` -> `x == !y` when the operands are of boolean types, or `!x` --> `x == TRUE`). I also fix a case where the fixpoint loop around the `&&` case would loop forever because `assign_rec` can both enlarge and shrink its value-set -- for example, the condition `x == 5 && **unknown == (void*)also_unknown` would result in repeatedly adding `x -> 5` and the clobbering the entire value set. I fix that by noting that in the context of two-way propagation, since there's no write operation taking place we should never shrink the value set.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: d61ce78738cb42e84e5cd3642de9444da224ea8a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78115950

File: src/analyses/constant_propagator.h
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/constant_propagator.h]: I'm not a fan of this being a public member. IMHO would be better to have this private. If this needs to be called from the outside, we could just call it `should_track_value_handler` and have a method `should_track_value` call this. It being public just invites shenanigans with people manually overwriting it.

File: src/analyses/constant_propagator.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/constant_propagator.cpp]: I think it'd be clearer to expand the not here: `lhs.is_constant() || !rhs.is_constant()`.

File: src/analyses/constant_propagator.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/constant_propagator.cpp]: `rhs.id() == ID_constant` should always be true at this point.

File: unit/analyses/constant_propagator.cpp
Comment (hannes-steffenhagen-diffblue) [inline unit/analyses/constant_propagator.cpp:194]: Would it be possible / reasonable to break this up into multiple smaller examples? Something this big is a bit difficult to follow.

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
I don't completely understand the two way propagation, but this looks reasonable to me.

**Comment by hannes-steffenhagen-diffblue** [comment]
Marking do not merge as it's based on another commit that should go in first

File: unit/analyses/constant_propagator.cpp
Comment (peterschrammel) [inline unit/analyses/constant_propagator.cpp]: This constructor is deprecated.

File: unit/analyses/constant_propagator.cpp
Comment (peterschrammel) [inline unit/analyses/constant_propagator.cpp]: This constructor is deprecated.

**Comment by smowton** [comment]
Comments mostly applied-- but this is parked for now pending #2132 landing.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: bf3d12c3e25cc89aa3f0b7d3bd1531902bcec5a5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78301619

File: src/analyses/constant_propagator.cpp
Comment (danpoe) [inline src/analyses/constant_propagator.cpp]: Could we use `assign_rec()` here?

**Review by danpoe — APPROVED** [review]
Looks good to me

**Comment by smowton** [comment]
Noting this is still blocked on 2132 as noted above, I believe action @tautschnig

**Comment by tautschnig** [comment]
> Noting this is still blocked on 2132 as noted above, I believe action @tautschnig

... which in turn is blocked by #2221 - @kroening @peterschrammel @smowton ;-)

**Comment by tautschnig** [comment]
In light of #2132 let me dare and go for high-level comments first:
1) It may be worth removing the changes to two-way-propagation from  #2132, and instead put this in to this PR here. Notably, that commit introduces a regression test that I'd like to have in place before any of the changes go in.
2) There are simple improvements in this PR that help readability a lot but don't even change behaviour, such as removing the return value. Maybe that could go in a commit of its own? No tests needed as it would be covered by the regression test.
3) Then there are strict extensions of functionality, such as dealing with negation or plain Booleans. Again, might be worth a commit of its own. That commit could nicely include tests showcasing the improvement.
4) Finally, there is the refactoring of existing functionality, which is probably the part that requires the most care. As such it would be really nice to have all the other parts factored out so that one can very easily read that bit of code/that particular change.
5) And then there's the test: unit tests are good, because they very selectively test functionality. But also they can be very hard to read, because they have to do a lot of work that is done by other components. As such I'd prefer to also have plain C examples. Now throwing away all the work that has gone into the unit test would also seem like a bad idea. Can we maybe have both?

**Comment by tautschnig** [comment]
@smowton #2132 is in, so this one should be ready to be rebased.

**Comment by smowton** [comment]
@tautschnig rebased and improved somewhat, and described the new state of affairs in the PR description -- please re-review

**Comment by tautschnig** [comment]
@smowton clang-format demands more work :-)

File: src/analyses/constant_propagator.cpp
Comment (tautschnig) [inline src/analyses/constant_propagator.cpp]: How about changing the above `else` to `else if(is_assignment)` and then even simplifying the comment a bit?

File: src/analyses/constant_propagator.cpp
Comment (tautschnig) [inline src/analyses/constant_propagator.cpp]: Unrelated change?

File: src/analyses/constant_propagator.cpp
Comment (tautschnig) [inline src/analyses/constant_propagator.cpp:329]: Would `std::any_of` work?

File: unit/analyses/constant_propagator.cpp
Comment (tautschnig) [inline unit/analyses/constant_propagator.cpp]: Use the new `code_ifthenelset` constructor.

File: unit/analyses/constant_propagator.cpp
Comment (tautschnig) [inline unit/analyses/constant_propagator.cpp]: There is a global one defined for the unit tests.

File: unit/analyses/constant_propagator.cpp
Comment (tautschnig) [inline unit/analyses/constant_propagator.cpp]: Might use the initializer-list constructor

File: unit/analyses/constant_propagator.cpp
Comment (tautschnig) [inline unit/analyses/constant_propagator.cpp]: Use recently added constructor

File: unit/analyses/constant_propagator.cpp
Comment (tautschnig) [inline unit/analyses/constant_propagator.cpp]: `move_to_operands` is deprecated

File: unit/analyses/constant_propagator.cpp
Comment (tautschnig) [inline unit/analyses/constant_propagator.cpp]: Use global one

**Review by tautschnig — APPROVED** [review]
Could do with a bunch of cleanup as indicated in comments, but otherwise ok.

File: src/analyses/constant_propagator.cpp
Comment (smowton) [reply src/analyses/constant_propagator.cpp:329]: No, as it would terminate early; we want to run propagate on every operand without regard for whether any individual sub-condition propagated anything.

**Comment by smowton** [comment]
@tautschnig @chrisr-diffblue @peterschrammel this is updated and ready for re-review.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 06e657352ef9c91611a36089c7bd2cdbba342700).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96099635

File: src/analyses/constant_propagator.cpp
Comment (tautschnig) [inline src/analyses/constant_propagator.cpp:352]: I'd use `expr.type().id() == ID_bool` to be consistent with what you're using above.

File: src/analyses/constant_propagator.cpp
Comment (tautschnig) [inline src/analyses/constant_propagator.cpp]: I'd rewrite this code a bit to do an early `if(lhs.id() != ID_typecast) return;` and then use `to_typecast_expr` instead of using the generic `.op0()`.

File: unit/analyses/constant_propagator.cpp
Comment (tautschnig) [inline unit/analyses/constant_propagator.cpp]: Does this function actually still add value? Doing this one-liner inline might aid clarity.

**Review by tautschnig — APPROVED** [review]

**Comment by smowton** [comment]
Updated as requested (except comparison with `bool_typet`, which I think is better than checking for `ID_bool`). Will merge when CI passes.

**Comment by tautschnig** [comment]
> (except comparison with `bool_typet`, which I think is better than checking for `ID_bool`)

I'm not disputing that, but I'm unhappy about mixing both approaches in a single commit. But this is a very minor issue indeed.

**Comment by smowton** [comment]
Ah sure, switched to use full type comparison everywhere.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 897ffdaf17df57b4b19eeb3944999a93cb575d8f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96195386

**Review by chrisr-diffblue — APPROVED** [review]
LGTM

**Pull request merged**
